tornado/testing.py

```diff
@@ -499,13 +506,16 @@ def wrap(f):
         @functools.wraps(f)
         def pre_coroutine(self, *args, **kwargs):
             result = f(self, *args, **kwargs)
-            if isinstance(result, GeneratorType):
+            if isinstance(result, GeneratorType) or iscoroutine(result):
                 self._test_generator = result
             else:
                 self._test_generator = None
             return result
 
-        coro = gen.coroutine(pre_coroutine)
+        if iscoroutinefunction(f):
```

Comment:
Why is this `iscoroutinefunction(f)` instead of `iscoroutine(result)`? We should prefer `isawaitable` to `iscoroutine*` whenever possible because it is compatible with cython on python 2.7 via [`backports_abc`](https://github.com/cython/backports_abc). 

I thought it would work to unconditionally wrap an `async def` function in `gen.coroutine`. Aren't coroutines a subtype of generators?